### PR TITLE
fix: prevent routing rule field bleed when rules are removed or reordered

### PR DESCRIPTION
**Pull request by sena1or**
## Problem

`threexui_xray_routing` corrupts its `rule` list whenever rules are removed or reordered: stale matchers from the prior rule at the same index bleed into the new rule. For example, reordering `[private→direct, RU-domains→direct, catch→proxy]` to `[RU-domains→direct, geoip:cn→direct, catch→proxy]` produces:

- `rule 0` (RU-domains) gains `ip:[geoip:private]` (stale, from the removed private rule)
- `rule 1` (geoip:cn) gains `network:"tcp,udp"` (stale, from the catch rule that used to be at index 1)

The merged rules are then written to the panel on apply, and read back into state — so every subsequent `terraform plan` shows \"No changes\" while the panel silently runs the wrong routing.

## Root cause

The nested `rule` block attributes are `Optional + Computed` with `UseStateForUnknown` (added by #228 to silence `(known after apply)` drift after import). `ListNestedBlock` elements are matched **by index**, so on reorder/removal the modifier copies the prior rule's unset fields into the new rule occupying the same index. This happens at plan time, polluting both what's written and (via read) the state.

## Fix

Add a resource-level `ModifyPlan` (`XrayRoutingResource` now implements `resource.ResourceWithModifyPlan`) that overrides the planned `rule` list with the configured one, so configuration is authoritative.

This is safe because:
- The panel echoes the routing template verbatim on save.
- The provider already filters the panel-managed `api` and `xui-dns-allow` rules on read (`isInternalAPIRoutingRule` / `isManagedDnsAllowRule`).
- So there are no legitimate computed-only routing fields — configuration is always authoritative and the override never drops a real value.

Unchanged rules (config == plan) are left untouched. Create (no prior state) and Delete (null plan) short-circuit before decoding.

The core logic lives in `reconcileRoutingPlan` for direct unit testability. The same `Optional+Computed+UseStateForUnknown` pattern exists on other `ListNestedBlock`s (`xray_outbounds`, `xray_dns`, `xray_basics`); this PR intentionally scopes the fix to `xray_routing` (the resource whose rules are routinely reordered/removed) per the minimal-change guideline. AGENTS.md documents the pattern and the fix.

## Tests

- **Unit** (`xray_routing_test.go`): `TestReconcileRoutingPlan_StripsStaleCarryForward`, `TestReconcileRoutingPlan_NoOpWhenUnchanged`, `TestXrayRoutingResource_ModifyPlan_StripsStaleCarryForward` (full framework path: decode plan/config → reconcile → set plan), `TestXrayRoutingResource_ModifyPlan_NoOpOnCreate`.
- **Acceptance** (`acc_xray_test.go`): `TestAccXrayRoutingRulesReorderNoFieldBleed` — Step 1 `[private→direct, RU-domains→blocked, catch→blocked]`, Step 2 `[RU-domains→blocked, geoip:cn→direct, catch→blocked]` (reorder + remove + add); asserts each rule carries only its configured matchers (no stale `ip` on the domain rule, no stale `network` on the ip rule) + an idempotency step.

`task pre-commit`, `task test:unit`, and the targeted `task test:acc` run all pass. Coverage: `reconcileRoutingPlan` 100%, `ModifyPlan` 81.8% (uncovered lines are the null-plan/Create and diagnostics-error guards).

### CODE REVIEW

File: provider/resource_xray_settings.go [hunk 1]
```diff
@@ -423,6 +425,60 @@ func (r *XrayRoutingResource) ImportState(ctx context.Context, _ resource.Import
...
+// so there are no legitimate computed-only routing fields: configuration is
+// always authoritative and this override never drops a real value.
+//
+// Create (no prior state) and Delete (null plan) have nothing to reconcile.
+func (r *XrayRoutingResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	if req.Plan.Raw.IsNull() || req.State.Raw.IsNull() {
+		return
+	}
+	var plan, config XrayRoutingModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
```
Comment (batonogov) [inline provider/resource_xray_settings.go:473]: 🔴 **Blocker:** `rule` is represented as `[]XrayRoutingRule` in `XrayRoutingModel`, so `req.Plan.Get`/`req.Config.Get` cannot decode an unknown value for the entire collection. Updating an existing resource with a computed `dynamic "rule"` therefore fails with a `Value Conversion Error` (`Path: rule`, `Suggested Type: basetypes.ListValue`). I reproduced this with a real `terraform plan -refresh=false` at this SHA; the same plan succeeds at the base SHA. Read `rule` directly as `types.List` (or another unknown-capable type), preserve unknown values, and reconcile only when the configured collection is known. Please add a regression test with non-null prior state and an unknown whole `rule` list. See also hashicorp/terraform-plugin-framework#1025.

File: provider/acc_xray_test.go [hunk 2]
```diff
@@ -130,6 +130,66 @@ func TestAccXrayRouting(t *testing.T) {
...
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.0.outbound_tag", "blocked"),
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.0.domain.0", "geosite:category-ru"),
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.0.domain.#", "1"),
+					resource.TestCheckNoResourceAttr("threexui_xray_routing.test", "rule.0.ip"),
+					// rule 1: geoip:cn→direct — must keep its ip and NOT inherit
+					// the stale network:"tcp,udp" from the prior index-1 rule.
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.1.outbound_tag", "direct"),
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.1.ip.0", "geoip:cn"),
+					resource.TestCheckResourceAttr("threexui_xray_routing.test", "rule.1.ip.#", "1"),
+					resource.TestCheckNoResourceAttr("threexui_xray_routing.test", "rule.1.network"),
```
Comment (batonogov) [inline provider/acc_xray_test.go]: 🔴 **Regression-test gap:** step 1 has `domain = geosite:category-ru` at index 1, while step 2 places `ip = geoip:cn` at index 1. The actual index-based carry-forward adds the stale `domain` to the new IP rule, but this assertion checks for an absent `network`; `network` was at index 2 in the prior state. The test can therefore pass while the remote payload contains both `domain` and `ip`. Assert that `rule.1.domain` is absent (preferably check every omitted matcher field and the raw remote payload), or rearrange the fixture so `network` really is at the old index 1. The unit fixture should use the same real state/index mapping.

**Review by batonogov — CHANGES_REQUESTED** [review]
## Review: changes requested

The overall direction is correct: making the configuration authoritative prevents index-based carry-forward when the `rule` list is fully known. However, two blocking cases must be addressed before merge:

1. `ModifyPlan` cannot handle an unknown value for the entire `rule` collection and breaks valid updates that use a computed `dynamic` block. This was reproduced with a real `terraform plan`; the same scenario plans successfully on the base SHA.
2. The acceptance fixture checks a different field from the one actually carried forward at index 1, so the test can remain green while the remote rule is corrupted.

Please add framework/acceptance regression coverage for an unknown collection and for the complete set of omitted matcher fields. The full CI run must also complete successfully.

File: provider/resource_xray_settings.go (hunk 1 above)
Comment (sena1or) [reply provider/resource_xray_settings.go:473]: Both blockers and the Codecov coverage gap are addressed in 733fc4e.

**This blocker (unknown collection):** `ModifyPlan` now reads `rule` as `types.List` via `req.Config.GetAttribute(ctx, resourcepath.Root("rule"), &ruleAttr)` before the typed decode. If `ruleAttr.IsUnknown()` (computed `dynamic` block), it returns early and defers to the schema plan modifiers. Added regression test `TestXrayRoutingResource_ModifyPlan_SkipsUnknownRuleCollection` with non-null prior state and an unknown whole `rule` list (`tftypes.NewValue(ruleListType, tftypes.UnknownValue)`), asserting the plan is unchanged and no diagnostics are raised.

**Blocker 2 (acc test assertions):** Corrected the carry-forward check from `rule.1.network` to `rule.1.domain` (the actual stale carry from step1[1]). Aligned unit + acc fixtures to the same real step1→step2 mapping. Added `testCheckRoutingRulesExact` which asserts EVERY omitted matcher per rule (handles flatmap list `.#` vs string bare-key representation). Added `testAccCheckXrayConfigRoutingRulesNoStaleFields` with a `threexui_xray_config` data source (`depends_on` the routing resource) to verify the panel DB itself carries no stale fields.

**Codecov:** `ModifyPlan` and `reconcileRoutingPlan` are now both at 100% statement coverage. Added 5 new unit tests covering every `HasError` guard (GetAttribute, Plan.Get, Config.Get), the unknown-collection skip, the no-op/reconcile-false branch, and the Create short-circuit.

CI workflows are pending approval (`action_required`) since this is a fork PR — they'll run once approved.

File: provider/acc_xray_test.go (hunk 2 above)
Comment (sena1or) [reply provider/acc_xray_test.go]: Fixed in 733fc4e. Corrected the assertion to check `rule.1.domain` (the actual stale carry from step1[1]). Rearranged both unit and acc fixtures to use the same real state/index mapping: step1 [ip:private→direct, domain:category-ru→blocked, network→blocked] → step2 [domain:category-ru→blocked, ip:geoip:cn→direct, network→blocked]. Added `testCheckRoutingRulesExact` to check every omitted matcher field per rule, and `testAccCheckXrayConfigRoutingRulesNoStaleFields` to verify the raw remote payload via a `threexui_xray_config` data source with `depends_on`.

**Review by batonogov — APPROVED** [review]
Re-reviewed at exact SHA `5ac17c011ced0775386b9e4caae5bfa4d3606e50`. Both blocking findings are resolved. `ModifyPlan` now reads the configured `rule` collection as `types.List` before typed decoding and safely preserves an unknown whole collection; I verified the update path with a real `terraform plan -refresh=false` against existing state and a computed `dynamic rule`, and also verified a known collection with an unknown matcher value. The reorder fixture now checks the actual stale `rule.1.domain`, aligns unit and acceptance fixtures with the real index mapping, asserts every omitted matcher, and verifies the panel payload. The targeted acceptance test passes against 3x-ui v3.6.0, while a selective mutant that leaves only the stale domain is correctly rejected. Local unit tests, build, vet, formatting, and diff checks pass. Codecov reports all modified coverable lines covered. Exact-SHA CI and the compatibility matrix are green: 23 completed checks, no pending or unsuccessful checks. The implementation is ready to merge.
